test/tools/d_do_test.d

```diff
@@ -479,28 +479,81 @@ bool collectExtraSources (in string input_dir, in string output_dir, in string[]
     return true;
 }
 
-// compare output string to reference string, but ignore places
-// marked by $n$ that contain compiler generated unique numbers
+// Compare output string to reference string, but ignore places marked by
+// special format specifiers that contain generated unique numbers or
+// platform specific values.
+// Recognized specifiers are:
+//  $n$ - Matches any digit [0-9]
+//  $h$ - Matches any hexdigit [0-9a-fA-F]
+// It is assumed that the format specifier is always 3 characters.
 bool compareOutput(string output, string refoutput)
 {
-    import std.ascii : digits;
-    import std.utf : byCodeUnit;
+    import std.ascii : digits, fullHexDigits;
     for ( ; ; )
     {
-        auto pos = refoutput.indexOf("$n$");
-        if (pos < 0)
-            return refoutput == output;
-        if (output.length < pos)
+        auto npos = refoutput.indexOf("$n$");
+        auto hpos = refoutput.indexOf("$h$");
+        if (npos >= 0 && npos < cast(size_t)hpos)
+        {
+            if (matchUpToPosition(output, refoutput, npos, digits))
+                continue;
             return false;
-        if (refoutput[0..pos] != output[0..pos])
+        }
+        if (hpos >= 0)
+        {
+            if (matchUpToPosition(output, refoutput, hpos, fullHexDigits))
+                continue;
             return false;
-        refoutput = refoutput[pos + 3 ..$];
-        output = output[pos..$];
-        auto p = output.byCodeUnit.countUntil!(e => !digits.canFind(e));
-        output = output[p..$];
+        }
+        return refoutput == output;
     }
 }
 
+bool matchUpToPosition(ref string output, ref string refoutput, size_t pos, string match)
+{
+    import std.utf : byCodeUnit;
+    if (output.length < pos)
+        return false;
+    if (refoutput[0..pos] != output[0..pos])
+        return false;
+    if (!match.canFind(output[pos]))
+        return false;
+    refoutput = refoutput[pos + 3 .. $];
+    output = output[pos..$];
```

Comment:
Nit: at least within the function your padding style could be the same (you use both here `pos..$` and `p .. $`).
D Style would be the later.

Reply:
Also this should be sth. like `output[min(pos + 1, output.length) .. $]` to avoid checking the first character twice.